woocommerce/woocommerce-ios
# HACK Week: Remove workaround for UIViewControllerRepresentable

Part of #12258


## Description

When migrating the Hub menu to the split view for tablet support, I learned that our workaround for updating the parent navigation items causes incorrect items when switching between views. This PR removes the workaround in preparation for the migration to split view. Specifically:
- Coupon list: replaced `EnhancedCouponListViewController` with a SwiftUI implementation. This includes migrating the handling for list selection, searching coupons and adding coupons to SwiftUI.
- Settings, Reviews, Review details: replace parent observer with SwiftUI wrapper.

## Testing instructions

1. Settings
- Log in to the app.
- Switch to Menu tab.
- Select Settings and navigate to any subview then back.
- Confirm that everything works as expected.

2. Reviews
- Log in to the app.
- Switch to Menu tab.
- Select Reviews and navigate to any existing reviews.
- Confirm that everything works as expected.

3. Review notification
- Log in to a test store on the app.
- In wp-admin, submit a review for a product.
- Update the [apns](https://github.com/woocommerce/woocommerce-ios/blob/4857351967722cd18909c7318174587f8e370dee/review.apns) file with your store ID and notification ID of the new review. You can get this ID by using proxyman and check for the latest item returned in the `GET /rest/v1.1/notifications` request.
- Enable notifications on your simulator and drag the apns file to the simulator - a notification should appear.
- Select the notification, and confirm that the review detail screen is displayed properly.

4. Coupons
- Log in to a test store with no coupon.
- Navigate to the Orders tab and select "+".
- Add a product then select Add Coupon.
- Confirm that an empty coupon list is presented. 
- Select "Add a coupon" and confirm navigation on the presented alert. 
- Confirm that you can be navigated to the empty coupon list.
- Select Add coupon, confirm that a bottom sheet for discount types is presented.
- Select a discount type, confirm that the coupon creation form is presented.
- Add details then submit the new coupon. Confirm that the creation form is then dismissed and the new coupon is displayed on the coupon list.
- Select the new coupon, the coupon detail screen should be displayed.
- Navigate back, select search icon. Confirm that the search screen displays the added coupons.
- Select a coupon, confirm that the detail screen for that coupon is displayed.
- Delete a coupon, confirm that the detail screen is dismissed, and a notice is displayed if the deletion is successful.

## Screenshots


Review notification:

https://github.com/woocommerce/woocommerce-ios/assets/5533851/08d2acfd-f7c8-44c8-9b4f-0a3bf3113868



---
